[JIT] Print better error when class attribute IValue conversion fails

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50255 [JIT] Print better error when class attribute IValue conversion fails**

**Summary**
TorchScript classes are copied attribute-by-attribute from a py::object into
a `jit::Object` in `toIValue`, which is called when copying objects from
Python into TorchScript. However, if an attribute of the class cannot be
converted, the error thrown is a standard pybind error that is hard to
act on.

This commit adds code to `toIValue` to convert each attribute to an
`IValue` inside a try-catch block, throwing a `cast_error` containing
the name of the attribute and the target type if the conversion fails.

**Test Plan**
This commit adds a unit test to `test_class_type.py`
based on the code in the issue that commit fixes.

**Fixes**
This commit fixes #46341.

Differential Revision: [D25854183](https://our.internmc.facebook.com/intern/diff/D25854183)

Commit:
[JIT] Print better error when class attribute IValue conversion fails

**Summary**
TorchScript classes are copied attribute-by-attribute from a py::object into
a `jit::Object` in `toIValue`, which is called when copying objects from
Python into TorchScript. However, if an attribute of the class cannot be
converted, the error thrown is a standard pybind error that is hard to
act on.

This commit adds code to `toIValue` to convert each attribute to an
`IValue` inside a try-catch block, throwing a `cast_error` containing
the name of the attribute and the target type if the conversion fails.

**Test Plan**
This commit adds a unit test to `test_class_type.py`
based on the code in the issue that commit fixes.

**Fixes**
This commit fixes #46341.

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/python/pybind_utils.cpp:
```diff
@@ -157,8 +157,16 @@ IValue toIValue(py::handle obj, const TypePtr& type, c10::optional<int32_t> N) {
+        try {
+          const auto& contained = py::getattr(obj, attrName.c_str());
+          userObj->setSlot(slot, toIValue(contained, attrType));
+        } catch (std::exception& e) {
```
Nice change! should we still append message of `e` to new exception for sake of providing user more context about the error?

SplitInfinity replied:
Done.

gmagogsfm commented on torch/csrc/jit/python/pybind_utils.cpp:
```diff
@@ -157,8 +157,16 @@ IValue toIValue(py::handle obj, const TypePtr& type, c10::optional<int32_t> N) {
+        } catch (std::exception& e) {
+          throw py::cast_error(c10::str(
+              "Could not cast attribute ",
+              attrName,
```
Nit:
Surround `attrName` with single quotes to make it clearer that it is a name of attribute.

SplitInfinity replied:
Done.

Commit:
Update on "[JIT] Print better error when class attribute IValue conversion fails"


**Summary**
TorchScript classes are copied attribute-by-attribute from a py::object into
a `jit::Object` in `toIValue`, which is called when copying objects from
Python into TorchScript. However, if an attribute of the class cannot be
converted, the error thrown is a standard pybind error that is hard to
act on.

This commit adds code to `toIValue` to convert each attribute to an
`IValue` inside a try-catch block, throwing a `cast_error` containing
the name of the attribute and the target type if the conversion fails.

**Test Plan**
This commit adds a unit test to `test_class_type.py`
based on the code in the issue that commit fixes.

**Fixes**
This commit fixes #46341.

[ghstack-poisoned]
